mgr/dashboard: carbon setup and ui shel

ivoalmeida opened the pull request:
<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>


https://github.com/ceph/ceph/assets/1071297/48527947-7eaa-4021-be9d-d57d66cba8e6

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.html:
```diff
@@ -1,427 +1,243 @@
+  </cds-header-global>
+</cds-header>
+<!-- ***************************** -->
+<!-- UPSTREAM LOGO BRAND TEMPLATE  -->
```
we don't need to call it upstream explicitly anywhere.
```suggestion
<!--  LOGO BRAND TEMPLATE  -->
```

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.scss:
```diff
@@ -188,95 +196,96 @@ $sidebar-width: 200px;
-
-  .cd-navbar-primary a:focus {
-    outline: none;
-  }
-
-  ngx-simplebar {
-    height: 100%;
-  }
+  // a.dropdown-toggle {
+  //   position: relative;
+
+  //   &::after {
+  //     border: 0;
+  //     content: '\f054';
+  //     font-family: 'ForkAwesome';
+  //     font-size: 1rem;
+  //     margin-top: 2px;
+  //     position: absolute;
+  //     right: 20px;
+  //     transition: transform 0.3s ease-in-out;
+  //   }
+
+  //   &[aria-expanded='true']::after {
+  //     transform: rotate(90deg);
+  //   }
+  // }
+
+  // ul ul a {
+  //   background: lighten(vv.$secondary, 10);
+  //   font-size: 1.1em !important;
+  //   padding-left: 35px !important;
+  // }
+
+  // .cd-navbar-primary a:focus {
+  //   outline: none;
+  // }
+
+  // ngx-simplebar {
+  //   height: 100%;
+  // }
```
I am guessing these codes will be removed by the end of this PR? or are you planning to do a later PR for doing these cleanups?

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/styles/carbon-overrides.scss:
instead of calling this `carbon-overrides.scss` why not `_carbon-defaults.scss`. We have a `_bootstrap-defaults.scss`. The idea would be to later replace it with carbon. WDYT?

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/styles/themes/_upstream.scss:
lets name the file `_default.scss` rather than `_upstream.scss`. Do we need a file `_overrides.scss` to override the theme or we can just override it in the `styles_overrides.scss` files when it comes to downstream?

nizamial09 reviewed (commented):
Thank you @ivoalmeida for taking care of this. Overall it looks good. I'll point out some issues which I saw while testing this.

I like the fact that when you toggle the navigation, it shows you icons rather than completely hiding it. But when you click on any items apart from the `Dashboard`, it does drop-down action and list whatever is inside, down. Though we won't visibly see those informations since the nav bar is now only accomodating the icons. Should we introduce icons to the child components? or should we just expand the navigation when the user drop-downs on an item? Also when we show only the icons, maybe we should show the dropdown icon too, to let the user know this icon will show more options below?

https://github.com/ceph/ceph/assets/71764184/bf0301a3-3055-4765-b5d2-13eb4d8b22fa

Couple more feedback on above item
1. All the parent level items like `Cluster`, `Multi-Cluster` etc shows the title when you hover over it. But it does not show it when you hover on `Dashboard` item.
2. The `title` on the child-level items displays the parent name, not the child name.


On the right side of every page, I see a scroll bar (both horizontal and vertical) which is always visible, even on empty state page. And on some pages, it doubles like below
![image](https://github.com/ceph/ceph/assets/71764184/5065ff3f-4784-4643-afef-eddb01b85cce)

If I hover over an icon in the right nav items, it focus on the icon as well as the outside area of it which looks a bit weird. One more thing I noticed there is that, if you want to toggle an item, you'll have to directly click on the icon itself and not on the other areas. otherwise it wouldn't open up anything which is particularly annoying for `Tasks and Notifications`, since thats one area everyone tries to click regularly.
![image](https://github.com/ceph/ceph/assets/71764184/10ab5dd7-89b6-4f1b-860a-a7ac861fb35b)

I can open many dropdown items at once which shouldn't be happening.
![image](https://github.com/ceph/ceph/assets/71764184/00d930a5-9182-4e1e-9746-6dbfe515cc70)


Can we choose a different body size and show full text?
![image](https://github.com/ceph/ceph/assets/71764184/d165d893-0709-44ce-b6b5-78125c3bd56a)

`Documentation` and `API` used to show an external icon to indicate this opens up the result in a new tab, that's not working anymore.
![image](https://github.com/ceph/ceph/assets/71764184/4ffd444d-fe59-4e12-8c82-3ff1a38125d9)

Related to the above one,it doesn't anymore opens the `Documentation` and `API Docs` in a new tab. it gets opened in the same tab.

Also the ceph logo used to navigate to the dashboard page (which was a recent fix by @afreen23 ) which is not working anymore in this change.

We don't need to show `Logged in user` text in the identity dropdown. Also the `Signed in user` entry should be unselectable.
![image](https://github.com/ceph/ceph/assets/71764184/c26ddbbe-3e54-43a7-9fe4-ce087b37ad65)

nizamial09 commented:
one more thing which I observed now - on the active item in the navigation sidebar, I see a left border on that selection with some blue color. can it match our primary accent?

ivoalmeida commented:
> Thank you @ivoalmeida for taking care of this. Overall it looks good. I'll point out some issues which I saw while testing this.
> 
> I like the fact that when you toggle the navigation, it shows you icons rather than completely hiding it. But when you click on any items apart from the `Dashboard`, it does drop-down action and list whatever is inside, down. Though we won't visibly see those informations since the nav bar is now only accomodating the icons. Should we introduce icons to the child components? or should we just expand the navigation when the user drop-downs on an item? Also when we show only the icons, maybe we should show the dropdown icon too, to let the user know this icon will show more options below?

For now I just made it work like it did before. We can have a separate discussion about the introduction of icons for every single navigation item and re-activate rail navigation.

>  scrnli_6_3_2024_1-26-01.PM.mp4 
> Couple more feedback on above item
> 
> 1. All the parent level items like `Cluster`, `Multi-Cluster` etc shows the title when you hover over it. But it does not show it when you hover on `Dashboard` item.

Fixed

> 2. The `title` on the child-level items displays the parent name, not the child name.

Fixed

> 
> On the right side of every page, I see a scroll bar (both horizontal and vertical) which is always visible, even on empty state page. And on some pages, it doubles like below ![image](https://private-user-images.githubusercontent.com/71764184/335989722-5065ff3f-4784-4643-afef-eddb01b85cce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgxMTgyMzEsIm5iZiI6MTcxODExNzkzMSwicGF0aCI6Ii83MTc2NDE4NC8zMzU5ODk3MjItNTA2NWZmM2YtNDc4NC00NjQzLWFmZWYtZWRkYjAxYjg1Y2NlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjExVDE0NTg1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBmNjRiOGUxOTQ0YjNjMjFkY2Q1Y2I4NWVjNzMzYzcxNzU3YTZkMDUwNDlkMDM5Mjc3M2MyOGJjZjZiOWZjNDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Zups-v5M-pdbkBUk8vl6pUUUsHv3Hv_oClyAO3m0HI0)
> 

Fixed

> If I hover over an icon in the right nav items, it focus on the icon as well as the outside area of it which looks a bit weird. One more thing I noticed there is that, if you want to toggle an item, you'll have to directly click on the icon itself and not on the other areas. otherwise it wouldn't open up anything which is particularly annoying for `Tasks and Notifications`, since thats one area everyone tries to click regularly. ![image](https://private-user-images.githubusercontent.com/71764184/335990555-10ab5dd7-89b6-4f1b-860a-a7ac861fb35b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgxMTgyMzEsIm5iZiI6MTcxODExNzkzMSwicGF0aCI6Ii83MTc2NDE4NC8zMzU5OTA1NTUtMTBhYjVkZDctODliNi00ZjFiLTg2MGEtYTdhYzg2MWZiMzViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjExVDE0NTg1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ3MGJmODFjOTMyN2Y0M2JiMGE1OTY3NmVkYTA3Yzc4N2FjNjEyM2NlZmY3YmMxZTIzYjk4NWVlYzc4MTZkMWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WJCVSafJlnPOeyAvjyoOB2ndbPVZSDOSAC8j0KoyEUI)
> 

This is a problem with the angular carbon component header-action. I removed it and added a carbon class directly and it works correctly now.

> I can open many dropdown items at once which shouldn't be happening. ![image](https://private-user-images.githubusercontent.com/71764184/335990872-00d930a5-9182-4e1e-9746-6dbfe515cc70.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgxMTgyMzEsIm5iZiI6MTcxODExNzkzMSwicGF0aCI6Ii83MTc2NDE4NC8zMzU5OTA4NzItMDBkOTMwYTUtOTE4Mi00ZTFlLTk3NDYtNmRiZmU1MTVjYzcwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjExVDE0NTg1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcyMjEwYTEzYzJiZDNmMjY1OTNmMDliYzBkZWE1NmY0NGRiZWUyODZiYmY3ZTRlYWYzNjExYzZiMmE1NmE2MWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ZNuqy1XLJImyIAZjdYS2oYesfYp5Pw0QbOsgL7SCFD8)
> 

Fixed. Same fix as previous complain

> Can we choose a different body size and show full text? ![image](https://private-user-images.githubusercontent.com/71764184/335991150-d165d893-0709-44ce-b6b5-78125c3bd56a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgxMTgyMzEsIm5iZiI6MTcxODExNzkzMSwicGF0aCI6Ii83MTc2NDE4NC8zMzU5OTExNTAtZDE2NWQ4OTMtMDcwOS00NGNlLWI2YjUtNzgxMjVjM2JkNTZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjExVDE0NTg1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiZGZiZDExNDg4YWJlMjgxNzM0ZmE0MGZhODlkMzE3ODRmZTY2OTU1YTlkZjMzMDkwZjM0OWI5YzY0MmU4MGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9hxzze3K_CbFtoWRnBfoV_8liiYdY1XgcYbcfuRiZgw)
> 

This is how the overflow-menu works but the documentation gives a workaround to use carbon's classes instead of the angular carbon component and now it works as expected.

> `Documentation` and `API` used to show an external icon to indicate this opens up the result in a new tab, that's not working anymore. ![image](https://private-user-images.githubusercontent.com/71764184/335991459-4ffd444d-fe59-4e12-8c82-3ff1a38125d9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgxMTgyMzEsIm5iZiI6MTcxODExNzkzMSwicGF0aCI6Ii83MTc2NDE4NC8zMzU5OTE0NTktNGZmZDQ0NGQtZmU1OS00ZTEyLThjODItM2ZmMWEzODEyNWQ5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjExVDE0NTg1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYxM2IwZmJiODIwMDI0NmFhMWJjMjY3YzIwNTkxOTliMzMxMDA3MWY1ZjgwZDNmMTAzMDhjZWEwNzhjMTk5YzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Nq0CIyJoztHoEo5ascEMvadEn7iEYaF8B9b-cZs2GcE)
> 

Fixed

> Related to the above one,it doesn't anymore opens the `Documentation` and `API Docs` in a new tab. it gets opened in the same tab.
> 

Fixed

> Also the ceph logo used to navigate to the dashboard page (which was a recent fix by @afreen23 ) which is not working anymore in this change.
> 

Fixed

> We don't need to show `Logged in user` text in the identity dropdown. Also the `Signed in user` entry should be unselectable. ![image](https://private-user-images.githubusercontent.com/71764184/335992776-c26ddbbe-3e54-43a7-9fe4-ce087b37ad65.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgxMTgyMzEsIm5iZiI6MTcxODExNzkzMSwicGF0aCI6Ii83MTc2NDE4NC8zMzU5OTI3NzYtYzI2ZGRiYmUtM2U1NC00M2E3LTlmZTQtY2UwODdiMzdhZDY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjExVDE0NTg1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc3OWJiMDZhNzMxYTBmZDE5OTRjYjA3ZjE4MWMyYTUyOWY3YWUxNzIwNmY0ODgxZmQ2ZjFkYzVhM2NhNzk5MWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5smIyaSYxK2fD_KSa1ydi73PWemUSj7HtPtri1n0n4M)

Fixed

ivoalmeida commented:
> one more thing which I observed now - on the active item in the navigation sidebar, I see a left border on that selection with some blue color. can it match our primary accent?

The default theme is the upstream one now so you should see it now.

nizamial09 reviewed (approved):
This looks really great and thank you for addressing my reviews @ivoalmeida.

nizamial09 commented:
jenkins test dashboard

nizamial09 commented:
jenkins test make check
